### PR TITLE
feat: rewrite the server in Rust

### PR DESCRIPTION
**Pull request by Avimitin**
The purpose of this PR is to rewrite the API server with Rust and redesign the database. I've tried `express` to write the server, but I don't like its design. Also I think it is better to separate the server and bot implementation.

 This PR is a draft for now, so you can review the modification before any further development.

## Request for Comments

1. In new design, I would like to let the server to do all the job. Let the bot be like a communication channel between packagers and database.

So in this new design, I will implement more API routes for bot. The bot just need to receive, validate, package the messages in group, and send them to the server. Server will handle those request and do related operation to database.

```mermaid
flowchart LR

A[Group] --> B[bot]
B --> A
B --> C[server]
C --> B
C --> D[Database]
```

## TODO

- [x] Use sqlite to replace the raw file database
- [ ] Use Rust to rewrite the API server
	- [x] route `/pkg`
	- [ ] route `/delete`
		- [x] Tg notification
		- [x] Auto merge
		- [x] Auto unmark
		- [x] Clean up related package
			- [x] Create relationship between packages
		- [x] Auto CC
		- [x] Test Case [Normal Delete]
		- [ ] Test Case [Package Relationship]
		- [ ] Test Case [Package chain unmark]
	- [ ] route `/add`
	- [x] merge message by time period
- [ ] Rewrite the telegram bot to fit the new database design

### CODE REVIEW

File: server/migrations/20221121091440_initialize.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
+-- Add migration script here
+CREATE TABLE IF NOT EXISTS packager(
+  tg_uid INT NOT NULL UNIQUE PRIMARY KEY,
+  alias TEXT NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS pkg (
+  id INT AUTO INCREMENT PRIMARY KEY,
+  name TEXT NOT NULL,
+  assignee INT,
```
Comment (XieJiSS) [inline server/migrations/20221121091440_initialize.sql]: We also need `assigned_at`

File: server/migrations/20221121091440_initialize.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  FOREIGN KEY(assignee) REFERENCES packager(tg_uid)
+);
+
+CREATE TABLE IF NOT EXISTS mark(
+  name TEXT NOT NULL,
+  mark_by INT,
+  comment TEXT,
```
Comment (XieJiSS) [inline server/migrations/20221121091440_initialize.sql]: We also need `marked_at`

File: server/src/routes.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+pub struct State {
+    pub db_conn: sqlx::SqlitePool,
+}
+
+type Data = actix_web::web::Data<State>;
+
+#[derive(serde::Serialize)]
+struct ErrorJsonResp<'m> {
+    msg: &'m str,
+    details: String,
```
Comment (XieJiSS) [inline server/src/routes.rs]: I'd prefer `detail` since `s` usually indicates an array

File: server/src/routes.rs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+#[get("/add")]
+pub(super) async fn add() -> HttpResponse {
+    HttpResponse::Ok().json(HashMap::from([("msg", "/add")]))
```
Comment (XieJiSS) [inline server/src/routes.rs]: Note that these routes are for maintaining marks, not assigning packages

**Review by XieJiSS — CHANGES_REQUESTED** [review]

File: server/migrations/20221121091440_initialize.sql (hunk 1 above)
Comment (Avimitin) [reply server/migrations/20221121091440_initialize.sql]: Implement in 20e478155e80a8a8df2896ee0b74cedf55744492

File: server/migrations/20221121091440_initialize.sql (hunk 2 above)
Comment (Avimitin) [reply server/migrations/20221121091440_initialize.sql]: Implement in 20e478155e80a8a8df2896ee0b74cedf55744492

File: server/src/routes.rs (hunk 3 above)
Comment (Avimitin) [reply server/src/routes.rs]: Fixed in 3decf66a67deceb6545ee74e7e294062d030c271

File: server/src/routes.rs (hunk 4 above)
Comment (Avimitin) [reply server/src/routes.rs]: I will keep the same implementation of the old API.

**Comment by Avimitin** [comment]
FYI: The implementation of route `/pkg` is done. And all tests for that route are also finished. Simply run `./tests/server_test.sh` can invoke the test.

**Comment by Avimitin** [comment]
FYI: auto merge and sending notification to telegram group are implemented. A new table `assignment` is added to separate the package information and package assignment storage.

**Comment by Avimitin** [comment]
FYI, auto unmark is done

**Comment by Avimitin** [comment]
FYI: message queue implemented.

**Comment by Avimitin** [comment]
Route /delete basically implemented.

**Comment by XieJiSS** [comment]
Can we add A/B tests to make sure that the old / new version of bots have consistent behaviors? IMO for the old version we can read the JSON database to understand how it reacts to the input.

**Comment by Avimitin** [comment]
> Can we add A/B tests to make sure that the old / new version of bots have consistent behaviors? IMO for the old version we can read the JSON database to understand how it reacts to the input.

I will try to figure out how to implement it.

**Comment by Avimitin** [comment]
It is a pity that I have no time continue this project. Closed now.

**Pull request closed without merging**